src/lua/LuaFileSystem.cpp

```diff
@@ -7,15 +7,107 @@
 #include "LuaObject.h"
 #include "Pi.h"
 
-/*
- * Interface: FileSystem
- *
- * A global table that provides access to the filesystem.
- *
- * This interface is protected. It can only be used by scripts in Pioneer's
- * data directory. Mods and scripts in the user directory that try to use it
- * will get a Lua error.
- */
+#include "core/StringUtils.h"
+
+struct SplitURI
+{
+	std::string		 full_path;
+	std::string_view scheme;
+	std::string_view rel_path;
+};
+
+SplitURI ParseURIArg(lua_State* l, int numArg)
+{
+	SplitURI res;
+	size_t path_size;
+	const char* path_str = luaL_checklstring(l, numArg, &path_size);
+
+	res.full_path = std::string(path_str, path_size);
+	size_t sep_pos = res.full_path.find_first_of("://");
+	if (sep_pos != std::string::npos)
+	{
+		res.scheme = std::string_view(path_str, sep_pos);
+		sep_pos += 3;
+		res.rel_path = std::string_view(path_str + sep_pos, path_size - sep_pos);
+	}
+	return res;
+}
+
+struct FileSource
+{
+	FileSystem::FileSource* source;
+	bool					is_read_write;
+};
+
+FileSource DetermineFileSource(std::string_view scheme)
+{
+	FileSource res;
+	if (compare_ci(scheme, "user"))
+	{
+		res.source = &FileSystem::userFiles;
+		res.is_read_write = true;
+	} else if (compare_ci(scheme, "data"))
+	{
+		res.source = &FileSystem::gameDataFiles;
+		res.is_read_write = false;
+	} else
+	{
+		res.source = nullptr;
+		res.is_read_write = false;
+	}
+	return res;
+}
+
+static lua_CFunction l_original_io_open = nullptr;
+
+void LuaFileSystem::register_raw_io_open_function(lua_CFunction open)
+{
+	l_original_io_open = open;
+}
+
+int LuaFileSystem::l_patched_io_open(lua_State* L)
+{
+	SplitURI split_uri = ParseURIArg(L, 1);
+	if (split_uri.scheme.empty())
+	{
+		return luaL_error(L, "'%s' does not have a valid scheme, must be user:// or data://", split_uri.full_path.c_str());
+	}
+
+	const FileSource fs = DetermineFileSource(split_uri.scheme);
+	if (nullptr == fs.source)
+	{
+		return luaL_error(L, "'%s' does not have a valid scheme, must be user:// or data://", split_uri.full_path.c_str());
+	}
+
+	const char *mode_cstr = luaL_checkstring(L, 2);
+	for (const char* c = mode_cstr; *c != 0; ++c )
+	{
+		if (*c != 'r')
+		{
+			if (!fs.is_read_write)
+			{
+				luaL_error(L, "'%s' does not support file operations that modify it, only things in the user folder do", split_uri.full_path.c_str());
+				return 0;
+			}
+			break;
+		}
+	}
```

Comment:
Because mode `r+` is read-write (and I'd prefer if a mode of `r` is implied if no argument is passed), this code should most likely be:
```suggestion
	std::string_view mode_str = LuaPull<std::string_view>(L, 2, "r");
	if (mode_str.find_first_of("aw+") != std::string_view::npos && !fs.is_read_write) {
		return luaL_error(L, "File '%s' is read-only.", split_uri.full_path.data());
	}
```